fix(ivy): don't cache render parent on LContainer to support ngUpgrade

pkozlowski-opensource opened the pull request.

pkozlowski-opensource committed:
refactor(ivy): simplify logic around canInsertNativeNode

Previously presence and type of a parent tNode was split among
canInsertNativeNode, canInsertNativeChildOfView and canInsertNativeChildOfElement.
This commit centralises the logic in canInsertNativeNode thus simplifying
the overall logic and making canInsertNativeChildOfElement trivial.

mary-poppins commented:
You can preview 14cd15e at https://pr28011-14cd15e.ngbuilds.io/.

pkozlowski-opensource committed:
fix(ivy): don't cache render parent on LContainer to support ngUpgrade

mary-poppins commented:
You can preview b1f91d1 at https://pr28011-b1f91d1.ngbuilds.io/.

alfaproject commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -569,19 +559,33 @@ function canInsertNativeChildOfView(viewTNode: TViewNode, view: LView): boolean
   } else {
-    // Parent is a regular element or a component
-    return canInsertNativeChildOfElement(currentNode);
+    assertNodeType(parent, TNodeType.Element);
```
Should this be a dev only assertion?

pkozlowski-opensource replied:
It should!  Good catch, fixing now!

pkozlowski-opensource committed:
refactor(ivy): remove concept of RENDER_PARENT from LContainer

mary-poppins commented:
You can preview 458dd1a at https://pr28011-458dd1a.ngbuilds.io/.

mary-poppins commented:
You can preview 6245f99 at https://pr28011-6245f99.ngbuilds.io/.

pkozlowski-opensource committed:
refactor(ivy): merge canInsertNativeNode and getParentNative

Previously the canInsertNativeNode and getRenderParent functions had almost
_exaclty_ the same logic. What was worse that getRenderParent was calling
canInsertNativeNode thus executing the same, non-trivial logic twice.

This commit merges canInsertNativeNode and getRenderParent into one function.
Now getRenderParent will return a native parent or null if a node can't be
inserted (content projection, root of a view that is not inserted etc.).

mary-poppins commented:
You can preview f6cb35c at https://pr28011-f6cb35c.ngbuilds.io/.

mary-poppins commented:
You can preview 6505d83 at https://pr28011-6505d83.ngbuilds.io/.

mary-poppins commented:
You can preview f4d3a5b at https://pr28011-f4d3a5b.ngbuilds.io/.

mhevery reviewed (approved):
❤️ it!

AndrewKushnir commented:
[Presubmit](http://test/OCL:228557406:BASE:228557436:1547062873425:c5429146)

kara commented on packages/upgrade/test/static/integration/upgrade_component_spec.ts:
```diff
@@ -2134,76 +2134,72 @@ withEachNg1Version(() => {
-                  expect(multiTrim(element.textContent)).toBe('ng2A(ng1(baz | ng2B(qux)))');
-                });
-              }));
+      it('should support single-slot transclusion', async(() => {
```
We should also remove the fixmes in `upgrade_spec` for these same cases:

https://github.com/angular/angular/blob/master/packages/upgrade/test/dynamic/upgrade_spec.ts#L2579

They are marked as FW-714, but I think that's outdated (based on desc in FW-863)

kara commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -561,30 +479,60 @@ function canInsertNativeChildOfView(viewTNode: TViewNode, view: LView): boolean
+    // We've got a parent which is an element in the current view. We just need to verify if the
+    // parent element is not a component. Component's content nodes are not inserted immediately
+    // because they will be projected, and so doing insert at this point would be wasteful.
+    // Since the projection would than move it to its final destination.
```
```suggestion
    // Since the projection would then move it to its final destination.
```

kara commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -561,30 +479,60 @@ function canInsertNativeChildOfView(viewTNode: TViewNode, view: LView): boolean
+      return getNativeByTNode(parent, currentView) as RElement;
+    } else {
+      return null;
+    }
```
Nit: I think this might be easier to read if we flip the conditional, so it's "if else" rather than "if not, else"  (ternary optional)

```ts
return parent.flags & TNodeFlags.isComponent ? null 
                                             : getNativeByTNode(parent, currentView) as RElement;
```

kara reviewed (approved):
I've got some nits, but they aren't blocking. Given time zones, we can probably just get this in and I'll fix it up in a follow-up?

kara commented:
See https://github.com/angular/angular/pull/28030
